### ISSUE TITLE
PPO rollouts not terminating with `done == True`

### ISSUE THREAD

**Issue by cisprague** [DLR-RM/stable-baselines3#23]
I am using a custom environment, and I've already checked the following:

```python
from stable_baselines3.common.env_checker import check_env

env = CustomEnv(arg1, ...)
# It will check your custom environment and output additional warnings if needed
check_env(env)
```

But the PPO algo is continuing to `step(action)` when `done == True` (when the state is no longer in the bounds).

This is how I am interfacing with the algorithm:

```python
class Agent:

    def __init__(self, environment, name, net_arch=[100, 100], n_env=1, n_steps=10000):

        # vectorise environment
        self.environment = environment
        check_env(self.environment)
        venv = DummyVecEnv([lambda: environment]*n_env)

        # load or create model
        assert(isinstance(name, str))
        self.name = name
        try:
            self.model = PPO.load(self.name, venv)
        except:
            self.model = PPO(
                'MlpPolicy', 
                venv,
                use_sde=True,
                sde_sample_freq=5,
                gae_lambda=0.9,
                learning_rate=1e-2,
                verbose=1,
                policy_kwargs=dict(net_arch=net_arch),
                n_steps=n_steps
            )

    def train(self, time_steps):

        # learn and save
        self.model.learn(total_timesteps=time_steps)
        self.model.save(self.name)

    def evaluate(self):

        # simulate
        obs = self.environment.reset()
        while True:
            action, _ = self.model.predict(obs, deterministic=True)
            obs, rew, done, _ = self.environment.step(action)
            if done: break

        # plot
        self.environment.system.plot(fname='{}.pdf'.format(self.name))
```

**Comment by Miffyli** [comment]
This is by design: PPO/A2C/etc gather samples for exactly `n_step` amount, crossing episode boundaries (i.e. continues from new episode once previous finishes).

**Comment by cisprague** [comment]
@Miffyli, thanks for the quick feedback. So, in this case, the reward itself should be enough to eventually enforce that I want the agent to stay within the bounds, right?

**Comment by araffin** [comment]
Wait, correct me if I'm wrong but you have an infinite loop in your `def evaluate(self):`.
And why are you not using `evaluate_policy` helper?

**Comment by cisprague** [comment]
@araffin, oops, yes, you're right, thanks! I've updated it now with `if done: break`, `evaluate` isn't used in training anyway. I haven't looked into `evaluate_policy` yet — my `evaluate` was for the purpose of plotting a trained trajectory.

**Comment by Miffyli** [comment]
>So, in this case, the reward itself should be enough to eventually enforce that I want the agent to stay within the bounds, right?

If I understood this right (you desire agent to avoid specific situations), then yes, a correct reward and/or terminal states should be able to teach agent to avoid those situations.

**Comment by cisprague** [comment]
@Miffyli, thanks. But, after more thought, I actually need the agent to stay within the state-space bounds during training, because some my underlying code requires that. How can I enforce this?

**Comment by Miffyli** [comment]
That escapes a bit outside stable-baselines's grasp. Something like [action space masking](https://github.com/hill-a/stable-baselines/issues/351) could possibly do it, but that is not implemented here. You might want to search up "safe reinforcement learning", where they discuss this kind of forbidden actions and such.

**Comment by cisprague** [comment]
@Miffyli, I mean I just want to terminate the episode once `done == True`, that's all.

**Comment by Miffyli** [comment]
Hmm that should be up to environment. Once agent receives `done == True`, it will call `env.reset()` next, thus resetting the environment. PPO algorithm keeps gathering these samples even though environment resets, up to the `n_step` points.

**Comment by cisprague** [comment]
Very odd, I restarted my computer and now it works properly now. Thanks for the help, you can delete this issue.
